src/gazebo_vision_plugin.cpp

```diff
@@ -113,21 +143,177 @@ void VisionPlugin::Load(physics::ModelPtr model, sdf::ElementPtr sdf)
+void VisionPlugin::handle_vision_data(double sec, Eigen::Vector3d &p, Eigen::Quaterniond &q, Eigen::Vector3d &lin_vel, Eigen::Vector3d &ang_vel, std::array<double, 36> &pose_cov, std::array<double, 36> &twist_cov)
+{
+   // Fill odom msg
+    odom_msg.set_usec(sec * 1e6); // convert sec to usec
+
+    gazebo::msgs::Vector3d* position = new gazebo::msgs::Vector3d();
+    position->set_x(p.x());
+    position->set_y(p.y());
+    position->set_z(p.z());
+    odom_msg.set_allocated_position(position);
+
+    gazebo::msgs::Quaternion* orientation = new gazebo::msgs::Quaternion();
+    orientation->set_x(q.x());
+    orientation->set_y(q.y());
+    orientation->set_z(q.z());
+    orientation->set_w(q.w());
+    odom_msg.set_allocated_orientation(orientation);
+
+    gazebo::msgs::Vector3d* linear_velocity = new gazebo::msgs::Vector3d();
+    linear_velocity->set_x(lin_vel.x());
+    linear_velocity->set_y(lin_vel.y());
+    linear_velocity->set_z(lin_vel.z());
+    odom_msg.set_allocated_linear_velocity(linear_velocity);
+
+    gazebo::msgs::Vector3d* angular_velocity = new gazebo::msgs::Vector3d();
+    angular_velocity->set_x(ang_vel.x());
+    angular_velocity->set_y(ang_vel.y());
+    angular_velocity->set_z(ang_vel.z());
+    odom_msg.set_allocated_angular_velocity(angular_velocity);
+
+    for (int i = 0; i < 36; i++) {
+      odom_msg.add_pose_covariance(pose_cov[i]);
+      odom_msg.add_twist_covariance(twist_cov[i]);
+    }
+
+    _last_pub_time = _current_time;
+
+    // publish odom msg
+    _pub_odom->Publish(odom_msg);
+}
+
+void VisionPlugin::rosPoseCovCallBack(const geometry_msgs::PoseWithCovarianceStampedConstPtr& odomMsg)
+{
+  double dt = calcTimeStep();
+
+  if (dt > 1.0 / _pub_rate) {
+
+    Eigen::Vector3d position(odomMsg->pose.pose.position.x, odomMsg->pose.pose.position.y, odomMsg->pose.pose.position.z);
+    Eigen::Quaterniond orientation(odomMsg->pose.pose.orientation.w, odomMsg->pose.pose.orientation.x, odomMsg->pose.pose.orientation.y, odomMsg->pose.pose.orientation.z);
+    Eigen::Vector3d lin_vel(0, 0, 0); // linear velocity is not provided by sensor
+    Eigen::Vector3d ang_vel(0, 0, 0); // angular velocity is not provided by sensor
+
+    std::array<double, 36> pose_covariance{};
+    std::array<double, 36> twist_covariance{}; // all values are zero
+    std::copy(odomMsg->pose.covariance.begin(), odomMsg->pose.covariance.end(), pose_covariance.begin());
+
+    handle_vision_data(odomMsg->header.stamp.toSec(),
+                       position,
+                       orientation,
+                       lin_vel,
+                       ang_vel,
+                       pose_covariance,
+                       twist_covariance);
+  }
+}
+
+void VisionPlugin::rosPoseCallBack(const geometry_msgs::PoseStampedConstPtr& odomMsg)
+{
+  double dt = calcTimeStep();
+
+  if (dt > 1.0 / _pub_rate) {
+
+    Eigen::Vector3d position(odomMsg->pose.position.x, odomMsg->pose.position.y, odomMsg->pose.position.z);
```

Comment:
Check `eigen_conversions`: http://docs.ros.org/groovy/api/eigen_conversions/html/namespacetf.html.

Reply:
`tf::vectorMsgToEigen()` and `tf::quaternionMsgToEigen()`.

Reply:
It's ok: `#include <eigen_conversions/eigen_msg.h>`

Reply:
Done.